flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategyTest.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.jobmanager.scheduler.CoLocationGroup;
+import org.apache.flink.runtime.jobmanager.scheduler.CoLocationGroupImpl;
+import org.apache.flink.runtime.jobmanager.scheduler.SlotSharingGroup;
+import org.apache.flink.runtime.scheduler.strategy.TestingSchedulingExecutionVertex;
+import org.apache.flink.runtime.scheduler.strategy.TestingSchedulingTopology;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.guava31.com.google.common.collect.Lists;
+import org.apache.flink.shaded.guava31.com.google.common.collect.Sets;
+
+import org.assertj.core.data.Offset;
+import org.junit.jupiter.api.Test;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link TaskBalancedPreferredSlotSharingStrategy}. */
+class TaskBalancedPreferredSlotSharingStrategyTest {
```

Comment:
One question for testing: is there any common testing for `TaskBalancedPreferredSlotSharingStrategyTest` and `LocalInputPreferredSlotSharingStrategyTest`? I see some test method names are same.

Is it possible to extract an `AbstractSlotSharingStrategyTest`, moving some common tests to there and these 2 classes extend `AbstractSlotSharingStrategyTest`?

Reply:
It sounds good to me~
Would you mind putting this part of the optimization in a Independent ticket after merging this feature ?

-In addition to extracting common code, it is also necessary to consider the upgrade issue of Junit4
-If there are common parts, it may add some complexity to the review process and update process.

Anyway, I think both options are good. Please let me know what's your opinion.

Reply:
> In addition to extracting common code, it is also necessary to consider the upgrade issue of Junit4

In order to simply the review of https://github.com/apache/flink/pull/23635 , I submit an hotfix PR to upgrade the `LocalInputPreferredSlotSharingStrategyTest` to junit5.

> If there are common parts, it may add some complexity to the review process and update process.

If it has some common part, we just review the different part. It will simply the review process. And it doesn't introduce any duplicated code.

Reply:
`LocalInputPreferredSlotSharingStrategyTest` is upgraded to junit5 in the master branch, please rebase master and go ahead in your free time, thanks~